### PR TITLE
Get symbol kind

### PR DESCRIPTION
**Pull request by krux02**
fixes https://github.com/nim-lang/Nim/issues/7366

when n is of type `NimNode`, then you can do the following:

`let symKind = n.symbol.kind`

### CODE REVIEW

**Comment by krux02** [comment]
I don't get it. What is the problem in travis? I couldn't find anything that failed in the log.

**Comment by krux02** [comment]
Something I would like to discuss here is the usefulness of the `NimSym` type itself. The only functions that accept a `NimSym` as argument, are the following:

```Nim
proc symbol=(n: NimNode; val: NimSym)
proc `$`(s: NimSym): string
proc `==`(a, b: NimSym): bool
proc getImpl(s: NimSym): NimNode
# and with this PR also this one
proc kind(s: NimSym): NimSymKind
```

`getImpl` and `kind` are the only "useful" functions here, and except from the getter `symbol` there is no way to get a NimSym. Both `genSym`and `bindSym` return a `NimNode` not a `NimSym`. both `getImpl` and `kind` could easily be replaced by functions that also accept a `NimNode` as argument and expect the node to be of kind `nnkSym`. Of coure would it be required to be rename `kind` to something else like `symKind` or `symbolKind`.

```Nim
let impl = n.symbol.getImpl
let kind = n.symbol.kind
# would become
let impl = n.getImpl
let kind = n.symKind
```

**Comment by Araq** [comment]
> Something I would like to discuss here is the usefulness of the NimSym type itself. 

Yeah, we should do without, I agree, good points!

**Comment by krux02** [comment]
Should I implement `getImpl` and `symKind` as proposed and tag following functions as deprecated, in this PR?

```
proc symbol(n: NimNode): NimSym
proc symbol=(n: NimNode; val: NimSym)
proc `$`(s: NimSym): string
proc `==`(a, b: NimSym): bool
proc getImpl(s: NimSym): NimNode
```

**Comment by krux02** [comment]
There is one thing I would like to discuss here, too. The NimIdent
type is already almost obsolete. These are the functions that are all
methods that take a NimIdent

```nim

# NimIdent result
proc toNimIdent*(s: string): NimIdent
proc ident*(n: NimNode): NimIdent

# NimIdent argument
proc `!`*(s: string): NimIdent
proc `$`*(i: NimIdent): string
proc `==`*(a, b: NimIdent): bool
proc `ident=`*(n: NimNode, val: NimIdent)
proc newIdentNode*(i: NimIdent): NimNode
proc newCall*(theProc: NimIdent, args: varargs[NimNode]): NimNode
proc nestList*(theProc: NimIdent, x: NimNode): NimNode
```

The only function that doesn't work with a NimNode is `nestList`. And
that one can easily be ported.
It is just ``proc `$`(n: NimNode)`` that relies on 
``proc `$`(i: NimIdent)`` for it's implementation. But apart from
that, `NimIdent` is ready for deprecation, too.

I think the best road to deprecate `NimIdent` is to rename `symName` that I just implemented to just `name`, and make it work on both identifiers and symbols.

File: lib/core/macros.nim [hunk 1]
```diff
@@ -195,9 +195,24 @@ proc kind*(n: NimNode): NimNodeKind {.magic: "NKind", noSideEffect.}
...
+  proc symName*(symbol: NimNode): string {.magic: "NSymName", noSideEffect.}
+  #  ## Unlike the result of `repr` the symbol name is not unique.
+  proc getImpl*(symbol: NimNode): NimNode {.magic: "GetImpl", noSideEffect.}
+    ## retrieve the implementation of `symbol`. `symbol` can be a
+    ## routine or a const.
+else:
+  proc symName*(n: NimNode): string =
+    $n.symbol
+  proc getImpl*(symbol: NimNode): NimNode =
+    symbol.symbol.getImpl
```
[the remark this answers is not in the document]
Comment (krux02) [reply lib/core/macros.nim]: yes they are, otherwise bootstrapping breaks.

[the remark this answers is not in the document]
Comment (dom96) [reply lib/core/macros.nim]: I see.

**Comment by Araq** [comment]
Looks excellent, needs a changelog entry and more documentation overall. Also check if the existing documentation is outdated with this change.

File: lib/core/macros.nim (hunk 1 above)
[the remark this answers is not in the document]
Comment (krux02) [reply lib/core/macros.nim]: well at least when the bootstrapping compiler is updated, the when statement with the else branch can completely disappear.

**Comment by krux02** [comment]
using `strVal=` for nkIdent nkSym does not work. For consistency I could make it work, but I don't think they should be used to set the string in either symbols not identifiers. Apart from that, this PR is ready for another review. A lot of changes went into this now.

**Comment by Araq** [comment]
It's really nice. I have nothing to complain. Not merging though since you had some questions in IRC.

**Pull request merged**